activesupport/lib/active_support/message_verifier.rb

```diff
@@ -46,16 +46,17 @@ def initialize(secret, options = {})
     def valid_message?(signed_message)
       return if signed_message.nil? || !signed_message.valid_encoding? || signed_message.blank?
 
-      data, digest = signed_message.split("--")
-      data.present? && digest.present? && ActiveSupport::SecurityUtils.secure_compare(digest, generate_digest(data))
+      parts = signed_message.split('.')
+      parts.size == 3 && parts.all?(&:present?) && untampered?(parts.pop, parts.join('.'))
     end
 
-    # Decodes the signed message using the +MessageVerifier+'s secret.
+    # Decodes the signed message using the +MessageVerifier+'s secret and returns the claims hash.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
     #
-    #   signed_message = verifier.generate 'a private message'
-    #   verifier.verified(signed_message) # => 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_at: 1.day.from_now)
+    #   verifier.verified(signed_message)
+    #   # => {:pld=>"a private message", :for=>"test", :exp=>"2015-06-29T07:16:12.579Z"}
```

Comment:
This breaks backwards compatibility. This should absolutely not change :smile:

I'm curious how you arrived at this as the acceptable return value? Perhaps there's something we haven't talked about enough :smile:

See, I view the JWT generation as an implementation detail that helps us support more options. My stance is that people shouldn't have to care about how we accomplish this.